crates/dojo-lang/src/interface.rs

```diff
@@ -0,0 +1,149 @@
+use cairo_lang_syntax::node::ast::{MaybeTraitBody, OptionReturnTypeClause};
+use cairo_lang_syntax::node::db::SyntaxGroup;
+use cairo_lang_utils::unordered_hash_map::UnorderedHashMap;
+use dojo_types::system::Dependency;
+
+use crate::plugin::{DojoAuxData, SystemAuxData};
+
+pub struct DojoInterface {
+    diagnostics: Vec<PluginDiagnostic>,
+    dependencies: HashMap<smol_str::SmolStr, Dependency>,
+}
+
+impl DojoInterface {
+    pub fn from_trait(db: &dyn SyntaxGroup, trait_ast: ast::ItemTrait) -> PluginResult {
+        let name = trait_ast.name(db).text(db);
+        let mut system = DojoInterface { diagnostics: vec![], dependencies: HashMap::new() };
+
+        if let MaybeTraitBody::Some(body) = trait_ast.body(db) {
+            let body_nodes: Vec<_> = body
+                .items(db)
+                .elements(db)
+                .iter()
+                .flat_map(|el| {
+                    if let ast::TraitItem::Function(fn_ast) = el {
+                        return system.rewrite_function(db, fn_ast.clone());
+                    }
+
+                    vec![RewriteNode::Copied(el.as_syntax_node())]
+                })
+                .collect();
+
+            let mut builder = PatchBuilder::new(db);
+            builder.add_modified(RewriteNode::interpolate_patched(
+                "
+                #[starknet::interface]
+                trait $name$<TContractState> {
+                    $body$
+                }
+                ",
+                &UnorderedHashMap::from([
+                    ("name".to_string(), RewriteNode::Text(name.to_string())),
+                    ("body".to_string(), RewriteNode::new_modified(body_nodes)),
+                ]),
+            ));
+
+            return PluginResult {
+                code: Some(PluginGeneratedFile {
+                    name: name.clone(),
+                    content: builder.code,
+                    aux_data: Some(DynGeneratedFileAuxData::new(DojoAuxData {
+                        models: vec![],
+                        systems: vec![SystemAuxData {
+                            name,
+                            dependencies: system.dependencies.values().cloned().collect(),
+                        }],
+                    })),
+                    code_mappings: builder.code_mappings,
+                }),
+                diagnostics: system.diagnostics,
+                remove_original_item: true,
+            };
+        }
+
+        PluginResult::default()
+    }
+
+    pub fn rewrite_function(
+        &mut self,
+        db: &dyn SyntaxGroup,
+        fn_ast: ast::TraitItemFunction,
+    ) -> Vec<RewriteNode> {
+        let mut rewrite_nodes = vec![];
+
+        // Some notes:
+        // * as rewritten functions belong to a starknet::interface:
+        //     - there is no generic parameter,
+        //     - there is no nopanic and implicits clauses.
+        // * there is no function body in a trait.
```

Comment:
Could we move this in a function documentation instead of notes inside the body?